gravitee-am-repository/gravitee-am-repository-jdbc/src/main/java/io/gravitee/am/repository/jdbc/management/api/JdbcDeviceRepository.java

```diff
@@ -48,39 +50,49 @@ public class JdbcDeviceRepository extends AbstractJdbcRepository implements Devi
     private static final String REF_TYPE_FIELD = "reference_type";
     private static final String ID_FIELD = "id";
     private static final String CLIENT_FIELD = "client";
-    private static final String USER_FIELD = "user_id";
+    private static final String USER_ID_FIELD = "user_id";
+    private static final String USER_EXTERNAL_ID_FIELD = "user_external_id";
+    private static final String USER_SOURCE_FIELD = "user_source";
     public static final String EXPIRES_AT_FIELD = "expires_at";
     public static final String DEVICE_IDENTIFIER_ID = "device_identifier_id";
     public static final String DEVICE_ID = "device_id";
 
-    protected Device toEntity(JdbcDevice entity) {
-        return mapper.map(entity, Device.class);
+    protected Device toEntity(JdbcDevice dbEntity) {
+        return dbEntity.toEntity();
     }
 
     @Override
-    public Flowable<Device> findByReferenceAndUser(ReferenceType referenceType, String referenceId, String user) {
+    public Flowable<Device> findByReferenceAndUser(ReferenceType referenceType, String referenceId, UserId userId) {
         LOGGER.debug("findByReferenceAndApplicationAndUser({}, {})", referenceType, referenceId);
         LocalDateTime now = LocalDateTime.now(UTC);
         return fluxToFlowable(getTemplate().select(JdbcDevice.class)
                 .matching(Query.query(
                         where(REFERENCE_ID_FIELD).is(referenceId)
                                 .and(where(REF_TYPE_FIELD).is(referenceType.name()))
-                                .and(where(USER_FIELD).is(user))
+                                .and(userMatches(userId))
                                 .and(where(EXPIRES_AT_FIELD).greaterThanOrEquals(now))
                 ))
                 .all())
                 .map(this::toEntity);
     }
 
+    private Criteria userMatches(UserId user) {
+        if (user.isExternal()) {
```

Comment:
```suggestion
        if (user.hasExternal()) {
```

Suggest to rename it `hasExternal` as we also check using the internalId in this case